pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -0,0 +1,530 @@
+	// expected matchedCache is prepared
+	if matchedCache == nil {
+		return framework.NewStatus(framework.Error, ErrReasonUnexpectedState)
+	}
+	return nil
+}
+
+func (p *Plugin) Score(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) (int64, *framework.Status) {
+	// if pod is a reserve pod, ignored
+	if IsReservePod(pod) {
+		return framework.MinNodeScore, nil
+	}
+
+	matchedCache := getPreFilterState(cycleState)
+	if matchedCache == nil { // expected matchedCache is prepared
+		return framework.MinNodeScore, framework.NewStatus(framework.Error, ErrReasonUnexpectedState)
+	}
+	rOnNode := matchedCache.GetOnNode(nodeName)
+	klog.V(4).InfoS("Attempting to score pod for reservation state", "pod", klog.KObj(pod),
+		"node", nodeName, "matched count", len(rOnNode))
+
+	// prefer the nodes which have reservation matched
+	if len(rOnNode) <= 0 {
+		return framework.MinNodeScore, nil
+	}
+	return framework.MaxNodeScore, nil
+}
+
+func (p *Plugin) ScoreExtensions() framework.ScoreExtensions {
+	return nil
+}
+
+func (p *Plugin) PreBind(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	// if the pod is a reserve pod
+	if IsReservePod(pod) {
+		return nil
+	}
+
+	// if the pod match a reservation
+	matchedCache := getPreFilterState(cycleState)
+	if matchedCache == nil { // expected matchedCache is prepared
+		return framework.NewStatus(framework.Error, ErrReasonUnexpectedState)
+	}
+	rOnNode := matchedCache.GetOnNode(nodeName)
+	if len(rOnNode) <= 0 { // the pod is suggested to bind on a node with no reservation
+		return nil
+	}
+
+	klog.V(3).InfoS("Attempting to pre-bind pod to node with reservations", "pod", klog.KObj(pod),
+		"node", nodeName, "matched count", len(rOnNode))
+	// select one reservation for the pod to allocate, update status
+
+	// sort: here we use MostAllocated (simply set all weights as 1.0)
+	for i := range rOnNode {
+		rOnNode[i].ScoreForPod(pod)
+	}
+	sort.Slice(rOnNode, func(i, j int) bool {
+		return rOnNode[i].Score >= rOnNode[j].Score
+	})
+	target := rOnNode[0].R
+
+	// update: update current owner and allocated resources info
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		curR, err1 := p.lister.Reservations(target.Namespace).Get(target.Name)
+		if errors.IsNotFound(err1) {
+			klog.Warningf("reservation %v not found, abort the update", GetReservationNamespacedName(target))
+			return nil
+		} else if err1 != nil {
+			klog.Warningf("failed to get reservation %v, err: %v", GetReservationNamespacedName(target), err1)
+			return err1
+		}
+
+		if !IsReservationScheduled(curR) {
+			klog.Warningf("failed to allocate resources on a non-scheduled reservation %v, phase %v",
+				GetReservationNamespacedName(curR), curR.Status.Phase)
+			// TBD: may try with the next candidate instead of return error
+			return fmt.Errorf(ErrReasonReservationExpired)
+		}
+
+		setReservationAllocated(curR, pod)
```

Comment:
before updating curR, please make DeepCopy... otherwise pollutes the data in the informer